sdmetrics/reports/single_table/diagnostic_report.py

```diff
@@ -0,0 +1,264 @@
+        ),
+        'WARNING': (
+            '! More than 10% the synthetic data does not follow the min/max boundaries set by '
+            'the real data'
+        ),
+        'DANGER': (
+            'x More than 50% the synthetic data does not follow the min/max boundaries set by '
+            'the real data'
+        ),
+    },
+    'CategoryCoverage': {
+        'SUCCESS': '✓ The synthetic data generally covers categories present in the real data',
+        'WARNING': (
+            '! The synthetic data is missing more than 10% of the categories present in the '
+            'real data'
+        ),
+        'DANGER': (
+            'x The synthetic data is missing more than 50% of the categories present in the '
+            'real data'
+        ),
+    },
+    'NewRowSynthesis': {
+        'SUCCESS': '✓ The synthetic rows are generally not copies of the real data',
+        'WARNING': '! More than 10% of the synthetic rows are copies of the real data',
+        'DANGER': 'x More than 50% of the synthetic rows are copies of the real data',
+    },
+    'RangeCoverage': {
+        'SUCCESS': (
+            '✓ The synthetic data generally covers numerical ranges present in the real data'
+        ),
+        'WARNING': (
+            '! The synthetic data is missing more than 10% of the numerical ranges present in '
+            'the real data'
+        ),
+        'DANGER': (
+            'x The synthetic data is missing more than 50% of the numerical ranges present in '
+            'the real data'
+        ),
+    }
+}
+
+
+class DiagnosticReport():
+    """Single table diagnostic report.
+
+    This class creates a diagnostic report for single-table data. It calculates the diagnostic
+    score along three properties - synthesis, coverage, and boundaries.
+    """
+
+    METRICS = {
+        'Coverage': [RangeCoverage, CategoryCoverage],
+        'Synthesis': [NewRowSynthesis],
+        'Boundaries': [BoundaryAdherence],
+    }
+
+    def __init__(self):
+        self._metric_results = {}
+        self._metric_averages = {}
+        self._results = {}
+
+    def _print_results(self, out=sys.stdout):
+        """Print the diagnostic report results."""
+        self._results['SUCCESS'] = []
+        self._results['WARNING'] = []
+        self._results['DANGER'] = []
+
+        for metric, score in self._metric_averages.items():
+            if np.isnan(score):
+                continue
+            if score >= 0.9:
+                self._results['SUCCESS'].append(RESULT_DETAILS[metric]['SUCCESS'])
+            elif score >= 0.5:
+                self._results['WARNING'].append(RESULT_DETAILS[metric]['WARNING'])
+            else:
+                self._results['DANGER'].append(RESULT_DETAILS[metric]['DANGER'])
+
+        out.write('DiagnosticResults:\n')
+        if len(self._results['SUCCESS']) > 0:
+            out.write('\nSUCCESS:\n')
+            for result in self._results['SUCCESS']:
+                out.write(f'{result}\n')
```

Comment:
minor but this could probably be pulled out into a function that takes in the out and the result level (warning, success...)